okhttp/src/main/java/okhttp3/Request.java

```diff
@@ -239,12 +262,29 @@ public Builder method(String method, @Nullable RequestBody body) {
       return this;
     }
 
+    /** Attaches {@code tag} to the request using {@code Object.class} as a key. */
+    public Builder tag(@Nullable Object tag) {
+      return tag(Object.class, tag);
+    }
+
     /**
-     * Attaches {@code tag} to the request. It can be used later to cancel the request. If the tag
-     * is unspecified or null, the request is canceled by using the request itself as the tag.
+     * Attaches {@code tag} to the request using {@code type} as a key. Tags can be read from a
+     * request using {@link Request#tag}. Use null to remove any existing tag assigned for {@code
+     * type}.
+     *
+     * <p>Use this API to attach timing, debugging, or other application data to a request so that
+     * you may read it in interceptors, event listeners, or callbacks.
      */
-    public Builder tag(Object tag) {
-      this.tag = tag;
+    public <T> Builder tag(Class<? extends T> type, @Nullable T tag) {
```

Comment:
My only complaint about this is that I can't easily store generic types or multiple of the same type. You can create wrapper or subclasses for statically known tagging, but you can dynamically add multiple. I don't know if that's something we care about probably being a `Map<String, Object>` where typing is the user problem.

Reply:
Main use cases I see

- oauth tokens
- tracing context flags
- logging tag set (zerorated, video, userinteractive, experimentlabels)

I think for any of these I don't expect that there will be hundreds of call sites in an app setting these on the request, so client dealing with String -> Object is ok.

Reply:
I’m imagining similar enough use cases, but I’m assuming boxing strings isn’t much of a hassle. I also think we could leverage existing types in tags:

```
Request request = ...
retrofit2.Call<?> retrofitCall = request.tag(retrofit2.Call.class);
zipkin.Span span = request.tag(zipkin.Span.class);
...
```

Reply:
I like that we can do this.. quite a lot.

Reply:
A tiny wrapper should help for generic type:
```java
interface MyListSupplier extends Supplier<List<String>> {}
...
List<String> myList = request.tag(ListSupplier.class).get();
...
```

Reply:
Assuming you have the flexibility to subtype and not wrap. And assuming your entries are statically known and not dynamic.

Reply:
Yup